documentation/PreferencesServer.md

```diff
@@ -349,6 +349,202 @@ There are two important things to note here:
+Also included is a `total_rows` and `offset` field, but do note that the `total_rows` field is not accurate as
+of the time of writing, and should be ignored.
+
+An example payload for a particlar safe may be:
+
+```json
+{
+    "total_rows": 2,
+    "offset": 0,
+    "rows": [{
+        "type": "gpiiCloudSafeCredential",
+        "etc etc": "..."
+    },
+    {
+        "type": "gpiiKey",
+        "etc etc": "..."
+    }]
+}
+```
+
+### POST /prefssafe-key-create
+
+This endpoint will add a new `gpiiKey` document to the system. There are no URL parameters, but the JSON body takes
+3 fields, one of them option. Required are the `prefsSafeId` and `prefsSetId`. These indicate the preferences safe,
+and the respective preferences set that this key and token will operate upon. Optionally you can pass in a unique
+unused `gpiiKey` to use, otherwise a new one will be generated as part of the process.  The new document is returned
+upon success. In event of a failure, a standard error document is returned with `isError` true, and a message
+detailing the failure.
+
+#### Example POST
+
+```json
+{
+    "prefsSafeId": "prefsSafe-1",
+    "prefsSetId": "gpii-default",
+    "gpiiKey": "3B3D3003-9F5F-4B66-98C1-1380EC86DDB1"
+}
+```
+
+Successful return payload:
+
+```json
+{
+    "type": "gpiiKey",
+    "schemaVersion": "0.1",
+    "prefsSafeId": "prefsSafe-1",
+    "prefsSetId": "gpii-default",
+    "revoked": false,
+    "revokedReason": null,
+    "timestampCreated": "2017-12-14T19:55:11.640Z",
+    "timestampUpdated": null
+}
+```
+
+### PUT /add-cloud-credential/:prefsSafeId
+
+This endpoint will add a new username and password credential set to the preferences safe whose id is specified
+in the URL parameter `prefsSafeId`. The body is a JSON payload containing the username and password for the new
+credential set.
+
+#### Example PUT
+
+URL: `http://preferences.gpii.net/add-cloud-credential/prefsSafe-1`
+
+The above indicates that the credentials will be added to preferences safe with `id` `prefsSafe-1`.
+
+putBody:
+
+```json
+{
+    "username": "NewLoginUsername",
+    "password": "v3ry$ecu4e"
+}
+```
+
+### POST /unlock-cloud-safe
+
+This endpoint allows you to verify if a username and password combination will unclock a preferences safe that
+has a matching cloud credential. It requires a username and password, but not a preferences safe ID.  If the
+unique combination unlocks a particular safe, that preferences safes `prefsSafe` document will be returned.
+
+#### Example post
+
+Using the `http://preferences.gpii.net/unlock-cloud-safe` with no parameters we can pass in the username and
+password via the body.
+
+postBody:
+
+```json
+{
+    "username": "NewLoginUsername",
+    "password": "v3ry$ecu4e"
+}
+```
+
+If successful this will return the preferences safe:
+
+```json
+{
+    "id": "prefsSafe-1",
+    "type": "prefsSafe",
+    "etc etc": "..."
+}
+```
+
+If the username and password do not match a record the following error is returned:

```

Comment:
If there could be different error responses, explain them all. Thanks.

Reply:
I put in some notes about error payloads for superfluous errors.